python/ray/_private/gcs_pubsub.py

```diff
@@ -139,51 +217,109 @@ def subscribe_error(self) -> None:
-                req = self._poll_request()
-                fut = self._stub.GcsSubscriberPoll.future(req, timeout=timeout)
-                # Wait for result to become available, or cancel if the
-                # subscriber has closed.
-                while True:
-                    try:
-                        fut.result(timeout=1)
+            # Make one polling request.
+            req = self._poll_request()
+            self._fut = self._stub.GcsSubscriberPoll.future(
+                req, timeout=timeout)
+            # Never return directly from this while loop. The cleanup after
+            # this while loop is necessary.
+            while True:
+                try:
+                    self._lock.release()
+                    # Cancel if the subscriber has closed.
+                    if self._close.is_set():
                         break
-                    except grpc.FutureTimeoutError:
-                        # Subscriber has closed. Cancel inflight the request
-                        # and return from polling.
-                        if self._close.is_set():
-                            fut.cancel()
-                            return None, None
-                        # GRPC has not replied, continue waiting.
-                        continue
-                    except Exception:
-                        # GRPC error, including deadline exceeded.
-                        raise
-                if fut.done():
-                    for msg in fut.result().pub_messages:
-                        self._errors.append((msg.key_id,
-                                             msg.error_info_message))
-
-            if len(self._errors) == 0:
-                return None, None
-            return self._errors.popleft()
+                    # Release lock and wait for at most 1s for result to
+                    # become available. This allows checking if the subscriber
+                    # has closed.
+                    self._fut.result(timeout=1)
+                    break
+                except grpc.FutureTimeoutError:
+                    # GRPC has not replied, continue waiting.
+                    continue
+                # TODO(mwtian): retry on connection failures.
+                except Exception:
+                    # This is GRPC error e.g. deadline exceeded. Abort polling.
+                    break
+                finally:
+                    # Re-acquire the lock before the next loop or before
+                    # exiting from the loop.
+                    self._lock.acquire()
+
+            # Process data from polling, or cancel the polling.
+            fut = self._fut
+            self._fut = None
+            self._cond.notify_all()
+            # Happens when subscriber is closed, or there is an exception
+            # during polling.
+            if fut.running():
+                fut.cancel()
+            # Also throws exception if the polling failed.
```

Comment:
The comment is wrong? Do we throw an exception here?

Reply:
Reworded to make it clear that it is intended that `fut.result()` throws exception on failure.